cpp/src/arrow/array.h

```diff
@@ -551,6 +551,17 @@ class BaseListArray : public Array {
   /// \brief Return array object containing the list's values
   std::shared_ptr<Array> values() const { return values_; }
 
+  /// \brief Return an Array that is a concatenation of the lists in this array.
+  ///
+  /// Note that it's different from `values()` in that it takes into
+  /// consideration of this array's offsets and slices its child Array
+  /// accordingly.
+  std::shared_ptr<Array> Flatten() const {
+    const offset_type begin_offset = value_offset(0);
+    const offset_type end_offset = value_offset(length());
+    return values()->Slice(begin_offset, end_offset - begin_offset);
```

Comment:
This doesn't accommodate nulls correctly. It's unknown whether the list "under" a null is empty or not. `ListBuilder` happens to produce `ListArray`s where this is the case, but it's not a guarantee. As an example of when this might happen, a new list array might be constructed from the child arrays of a list array with a new null bitmap.

You can use `Concatenate` to eat around the nulls.

Reply:
You are right.. But on the other hand I think it's important to have a O(1) flatten for those ListArrays whose empty sub-lists are behind the "null" elements (is that also the common case?).

The generic solution is O(len(ListArray)) as it needs to scan the null bitmap and the offsets the determine whether a concatenation is needed..

Do we want to consider introducing an indicator, or (if possible) just enforcing that empty lists should be behind nulls?